libraries/InternalFileSytem/src/flash/flash_nrf5x.c

```diff
@@ -140,27 +185,39 @@ static uint32_t fal_program (uint32_t dst, void const * src, uint32_t len)
   // https://devzone.nordicsemi.com/f/nordic-q-a/40088/sd_flash_write-cause-nrf_fault_id_sd_assert
   // Workaround: write half page at a time.
 #if NRF52832_XXAA
-  while ( NRF_ERROR_BUSY == (err = sd_flash_write((uint32_t*) dst, (uint32_t const *) src, len/4)) )
-  {
-    delay(1);
+  // Write the page
+  // Multiple attempts, if needed
+  for (uint8_t attempt = 0; attempt < MAX_RETRY; ++attempt) {
+    err = sd_flash_write((uint32_t*) dst, (uint32_t const *) src, len/4);
+
+    if(sd_en) err = wait_for_async_flash_op_completion(err); // Only async if soft device enabled
```

Comment:
the logic appears sound.  I do recommend always using curly braces around if..else, do..while, and while blocks.   (I earned this knowledge via pain.)   Even if you keep things on one line, use the curly braces.  It avoids weird things happening when someone later uses a macro, and that macro wasn't written to be single-statement safe.

Applies also twice more below.